src/alr/alr-commands-update.adb

```diff
@@ -1,42 +1,49 @@
-with Alire.Paths;
+with Alire.Containers;
+with Alire.Errors;
 with Alire.Solutions.Diffs;
 with Alire.Solver;
+with Alire.Utils.TTY;
+with Alire.Workspaces;
 
 with Alr.Checkout;
+with Alr.Commands.Index;
 with Alr.Commands.User_Input;
 with Alr.Platform;
 with Alr.Root;
 
-with GNAT.OS_Lib;
-with Alr.Bootstrap;
-
 package body Alr.Commands.Update is
 
-   use all type Bootstrap.Session_States;
-
    package Query renames Alire.Solver;
 
    -------------
    -- Upgrade --
    -------------
 
    procedure Upgrade (Interactive : Boolean;
-                      Force       : Boolean := False) is
-      --  The part concerning only to the working release
+                      Force       : Boolean := False;
+                      Allowed     : Alire.Containers.Crate_Name_Sets.Set :=
+                        Alire.Containers.Crate_Name_Sets.Empty_Set)
+   is
+      Old     : constant Query.Solution :=
+                  Root.Current.Solution;
    begin
-      Requires_Full_Index;
 
-      Requires_Valid_Session;
+      --  Ensure requested crates are in solution first
+
+      for Crate of Allowed loop
+         if not Old.Releases.Contains (Crate) then
+            Reportaise_Wrong_Arguments ("Requested crate is not in solution: "
```

Comment:
I feel like it would be more meaningful for the user to use `Requested crate is not a dependency` here. `Solution` is a bit internal vocabulary.

Reply:
D'accord